File: generated_code/python/solution_572_2.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        """
        Time Complexity: O(N * M) in the worst case, where N is the number of nodes 
        in 'root' and M is the number of nodes in 'subRoot'.
        Space Complexity: O(H_root) for the recursion stack, where H_root is the 
        height of the 'root' tree.
        """
        # If the main tree is empty, it cannot contain any non-empty subtree.
        # (Given constraints say subRoot has at least 1 node).
        if not root:
            return False
        
        # Check if the tree rooted at the current 'root' is identical to 'subRoot'
        if self.isSameTree(root, subRoot):
            return True
        
        # Otherwise, recursively check the left and right subtrees
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # If both nodes are null, they are identical
        if not p and not q:
            return True
        # If one is null and the other isn't, or their values differ, they aren't identical
        if not p or not q or p.val != q.val:
            return False
        
        # Recursively check if left subtrees and right subtrees are identical
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: generated_code/cpp/solution_572_2.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // Helper function to check if two trees are completely identical
    bool isSameTree(TreeNode* p, TreeNode* q) {
    // Both are null, so they are identical
    if (!p && !q) return true;
    // One is null and the other isn't, so they aren't identical
    if (!p || !q) return false;
        // Check if current values match and structurally identical on left and right subtrees
        return (p->val == q->val) && isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        // A null tree cannot contain any subtree (given subRoot has at least 1 node per constraints)
        if (!root) return false;
        
        // If the tree rooted at 'root' is identical to 'subRoot', we found it
        if (isSameTree(root, subRoot)) return true;
        
        // Otherwise, recursively check if 'subRoot' is a subtree of the left or right child
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }

};